packages/dds/tree/src/feature-libraries/persisted-config/persistedConfig.ts

```diff
@@ -0,0 +1,336 @@
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+// TODO: Document, decide if semver-like even provides value here (over e.g. plain-old number).
+export type Version = `${number}.${number}.${number}${"-internal" | ""}`;
+
+export interface PersistedConfig {
+	version: Version;
+	// Strictly increasing number which identifies the iteration of the proposed protocol. It's suggested
+	// that applications are able to modify this using the same mechanisms with which they can modify a DDS's
+	// configuration, which enables safe rollback policies.
+	// Example: application enables feature flag 1 in iteration 2, then realizes there is a problem with it.
+	// application then disables feature flag 1 in iteration 3.
+	// Using a single version number and monotonic increase on it isn't sufficient: DDSes may support configuration
+	// which impacts their serialization format with independent semantics (ex: attribution, policies to include history)
+	protocolIteration: number;
+	flags: Record<string, boolean>; // TODO: consider making this generic
+}
+
+export enum ConfigUpgradeType {
+	ConcurrentOpsInvalid,
+	ConcurrentOpsValid,
+}
+
+// note: would be great to lock this down. changing this for flags that have been added can have nasty side effects
+// for in progress upgrades.
+// TODO: document process for adding flags
+export interface PersistedConfigSchema {
+	version: (current: Version, previous: Version) => ConfigUpgradeType;
+	flags: Record<string, (current: boolean, previous: boolean) => ConfigUpgradeType>;
+}
+
+interface VersionControllerSummary {
+	config: Omit<PersistedConfig, "flags"> & Partial<Pick<PersistedConfig, "flags">>;
+	mostRecentResubmissionSeq?: number;
+}
+
+/**
+ * TODO: Better doc here. But the gist of it is typical use case would be a DDS that owns one of these.
+ * The structure of DDS APIs and how this interacts with them makes the usage contract relatively brittle currently--there are
+ * a few options for improving that. Still, this seems like an improvement to hard-coding the equivalent policy all over the place
+ * with similar settings.
+ * - Rather than call `submitLocalMessage`, they always use this object's `submit`.
+ * - DDS op processing logic should always defer to this object first, and avoid handling the op if `tryProcessOp` returns true.
+ * - DDS should include summary of the persisted config (obtained via `summarize`) somewhere in their summary
+ * - DDS should rehydrate persisted config state in their `loadCore` (i.e. retrieve from wherever they put it in
+ * 	the summary and call `loadCore` on this object)
+ * - When encoding op contents for submission, DDS should use configuration dictated by `getConfigForNextSubmission`.
+ *
+ * Note: may want to provide simpler API for users which don't need synchronization points, which means op resubmission is never required.
+ */
+export interface IPersistedConfigStore {
```

Comment:
The methods on this class generally seem like they could be baked higher up into the runtime so that they happen automatically and the DDS author barely has to worry about them (which I know is the ideal scenario).

As a middle ground, could you perhaps prototype a subclass of SharedTreeCore which hides as much of this as possible? E.g. in its implementation of `process`, it only calls `processCore` if `tryProcessOp` returns false, etc. Then a DDS wanting to use this for initial prototypes can just subclass that and hide away as much boilerplate as possible.

Reply:
subclassing is an option for usage, but note that it won't be totally transparent: invocation order of super.processCore and processCore isn't what you want (it's up to SharedTree processCore whether that gets called)

Reply:
I meant to put the change in `process`, and leave `processCore` alone. I just noticed that `process` is private, though. So it'd have to be a modification to `SharedObjectCore` directly. Which doesn't sound like a bad idea.